framework/src/modules/chain/submodules/blocks/process.js

```diff
@@ -427,11 +435,242 @@ class Process {
+	 * @return {Promise}
+	 * @private
+	 */
+	async _forkChoiceTask(block, newBlockReceivedAt) {
+		const lastBlock = modules.blocks.lastBlock.get();
+
+		const forgingSlotLastBlock = slots.getSlotNumber(lastBlock.timestamp);
+		const forgingSlotNewBlock = slots.getSlotNumber(block.timestamp);
+
+		// Slot number when lastBlock was received.
+		const lastBlockReceivedAt =
+			modules.blocks.lastReceipt.get() || lastBlock.timestamp;
+
+		if (lastBlock.id === block.id) {
+			// Case 1: same block received twice
+			return this._handleSameBlockReceived(block);
+		}
+
+		if (
+			lastBlock.height + 1 === block.height &&
+			block.previousBlock === lastBlock.id
+		) {
+			// Case 2: correct block received
+			return this._handleGoodBlock(block);
+		}
+
+		if (
+			lastBlock.height === block.height &&
+			lastBlock.heightPrevoted === block.heightPrevoted &&
+			lastBlock.previousBlock === block.previousBlock
+		) {
+			// Delegates are the same
+			if (lastBlock.generatorPublicKey === block.generatorPublicKey) {
+				// Case 3: double forging different blocks in the same slot.
+				// Last Block stands.
+				return this._handleDoubleForging(block);
+			}
+
+			// Two competing blocks by different delegates at the same height.
+			if (
+				forgingSlotLastBlock < forgingSlotNewBlock &&
+				!this._receivedInSlot(lastBlock, lastBlockReceivedAt) &&
+				this._receivedInSlot(block, newBlockReceivedAt)
+			) {
+				// Case 4: Tie break
+				return this._handleDoubleForgingTieBreak(block, lastBlock);
+			}
+		} else if (
+			lastBlock.heightPrevoted < block.heightPrevoted ||
+			(lastBlock.height < block.height &&
+				lastBlock.heightPrevoted === block.heightPrevoted)
+		) {
+			// Case 5: received block has priority. Move to a different chain.
+
+			return this._handleMovingToDifferentChain();
+			// TODO: Move to a different chain
+		}
+
+		// Discard newly received block
+		return library.logger.warn(
+			`Discarded block that does not match with current chain: ${
+				block.id
+			} height: ${block.height} round: ${slots.calcRound(
+				block.height
+			)} slot: ${slots.getSlotNumber(block.timestamp)} generator: ${
+				block.generatorPublicKey
+			}`
+		);
+	}
+
+	/**
+	 * Block IDs are the same ~ Blocks are equal
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleSameBlockReceived(block) {
+		library.logger.debug('Block already processed', block.id);
+	}
+
+	/**
+	 * Block received is correct
+	 * @param block
+	 * @returns {Promise}
+	 * @private
+	 */
+	_handleGoodBlock(block) {
+		return this._processBlock(block);
+	}
+
+	/**
+	 * Double forging. Last block stands
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDoubleForging(block) {
+		library.logger.warn(
+			'Delegate forging on multiple nodes',
+			block.generatorPublicKey
+		);
+		library.logger.info('Last block stands');
```

Comment:
Please elaborate this log message further along with the ids of both blocks.

Reply:
I have changed the log so it includes both block ids and a clear message.